### PR TITLE
[DNM] West manifest imports documentation

### PR DESCRIPTION
**Pull request by mbolivar**
Document manifest imports, a feature which will be in west 0.7 when it is released.

The rendered HTML for this documentation PR is here:

https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/20433/docs/guides/west/manifest.html#manifest-imports

Feedback is requested!

--------

TODO:

- [x] remove "(PLANNED)" markers from docs once map implementation is done
- [x] document related west API changes for 0.7 once that release is out

### CODE REVIEW

File: doc/guides/west/manifest.rst [hunk 1]
```diff
@@ -775,63 +772,62 @@ hosted at ``https://git.example.com/mainline/manifest``.
...
        - name: app
          url: https://git.example.com/downstream/app
-       - name: library
-         url: https://git.example.com/downstream/library
+       - name: lib3
+         path: libraries/lib3
+         url: https://git.example.com/downstream/lib3
 
-If a whitelist had not been used:
+If a whitelist had not been used, the ``lib2`` project from the mainline
```
Comment (tejlmand) [inline doc/guides/west/manifest.rst]: Something is wrong here.
Cause you actually do import `lib2` as it is in the whitelist.
```
           name-whitelist:
             - app
             - lib
```
did you mean `lib` instead of `lib2` ?

File: doc/guides/west/manifest.rst [hunk 2]
```diff
@@ -775,63 +772,62 @@ hosted at ``https://git.example.com/mainline/manifest``.
...
+manifest would have been imported.
 
-- The ``library`` project in the final manifest would have had its ``revision``
-  set to ``refs/heads/only-in-mainline``.
-- ``unnecessary-project`` would have been imported.
+If a rename had not been used, the downstream manifest's ``app`` project would
+have shadowed the mainline manifest's project by the same name.
```
Comment (tejlmand) [inline doc/guides/west/manifest.rst]: Is it clear that `rename` also changes the path of the imported project ?
And what would have happened in those cases ?:
```
# mainline west.yml:
manifest:
  projects:
    - name: app
      path: app-path
      url: https://git.example.com/mainline/app

# downstream west.yml:
manifest:
  projects:
    - name: mainline
      url: https://git.example.com/mainline/manifest
      import:
        name-whitelist:
          - app
          - lib2
        rename:
          app: mainline-app
    - name: app
      url: https://git.example.com/downstream/app
```
Will `mainline-app` be placed in path `app-path`, or in path `mainline-app`.

or this case, where both projects are defined with colliding paths.
```
# mainline west.yml:
manifest:
  projects:
    - name: app
      path: app-colliding-path
      url: https://git.example.com/mainline/app

# downstream west.yml:
manifest:
  projects:
    - name: mainline
      url: https://git.example.com/mainline/manifest
      import:
        name-whitelist:
          - app
          - lib2
        rename:
          app: mainline-app
    - name: app
      path: app-colliding-path
      url: https://git.example.com/downstream/app
```

File: doc/guides/west/manifest.rst [hunk 3]
```diff
@@ -840,44 +836,50 @@ using ``glob`` syntax.
      projects:
        - name: app
          url: https://git.example.com/mainline/app
-       - name: library
-         url: https://git.example.com/mainline/library
-         revision: refs/heads/only-in-mainline
-       - name: library2
-         url: https://git.example.com/mainline/library-2
-       - name: unnecessary-project
-         url: https://git.example.com/mainline/unnecessary-project
```
Comment (tejlmand) [inline doc/guides/west/manifest.rst]: Why is this removed ?
Without this, I don't think this example is useful, as nothing is exluded during import.

File: doc/guides/west/manifest.rst [hunk 4]
```diff
@@ -735,33 +735,30 @@ The ``import`` key can also contain a mapping with the following keys:
 
 - ``file``: The name of the manifest file or directory. This defaults
```
Comment (tejlmand) [inline doc/guides/west/manifest.rst]: I guess, `file` is also optional.
And if not provided, `west.yml` is used.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Found some minor issues with the examples.

File: doc/guides/west/manifest.rst (hunk 2 above)
Comment (tejlmand) [reply doc/guides/west/manifest.rst]: > Is it clear that rename also changes the path of the imported project ?

Looking at code, it seems path is not renamed, if set in mainline manifest, is that correct ?
And in that case, what happens if there is a path collision ?

**Comment by mbolivar** [comment]
kicking CI

File: doc/guides/west/manifest.rst (hunk 4 above)
Comment (mbolivar) [reply doc/guides/west/manifest.rst]: > I guess, file is also optional.

I'll add the word Optional, thanks.

> And if not provided, west.yml is used.

That's what it says on this line and the next one, no?

"This defaults to west.yml if not present"

File: doc/guides/west/manifest.rst (hunk 1 above)
Comment (mbolivar) [reply doc/guides/west/manifest.rst]: Eagle eye. Thanks for catching that. I did mean `lib`.

File: doc/guides/west/manifest.rst (hunk 2 above)
Comment (mbolivar) [reply doc/guides/west/manifest.rst]: I think I am just going to remove support for `rename` for now.

File: doc/guides/west/manifest.rst (hunk 3 above)
Comment (mbolivar) [reply doc/guides/west/manifest.rst]: The `lib` project is excluded, no?

**Comment by mbolivar** [comment]
@tejlmand thanks for review! Comments addressed.

File: doc/guides/west/manifest.rst (hunk 3 above)
Comment (tejlmand) [reply doc/guides/west/manifest.rst]: This is the example manifest:
```
# mainline west.yml:
   manifest:
     projects:
       - name: app
         url: https://git.example.com/mainline/app
      - name: lib
        path: libraries/lib
        url: https://git.example.com/mainline/lib
       - name: lib2
         path: libraries/lib2
         url: https://git.example.com/mainline/lib2

  # downstream west.yml:
   manifest:
     projects:
       - name: mainline
         url: https://git.example.com/mainline/manifest
         import:
           path-whitelist: libraries/*
```

and as both `lib` and `lib2` are located under the folder `libraries` and that the path `libraries` is given in `path-whitelist: libraries/*`, then both of them are included, and nothing is excluded. 

Therefore, to show a project being excluded in this example, you'll need the `unnecessary-project`

File: doc/guides/west/manifest.rst (hunk 4 above)
Comment (tejlmand) [reply doc/guides/west/manifest.rst]: > > And if not provided, west.yml is used.
> 
> That's what it says on this line and the next one, no?
> 
> "This defaults to west.yml if not present"

Yep, I only referenced that line, to show that you indirectly stated that it was optional.
As something you had already written.

File: doc/guides/west/manifest.rst (hunk 3 above)
Comment (tejlmand) [reply doc/guides/west/manifest.rst]: Just noticed. Project `app` is actually excluded.
And in that case `unnecessary-project` is not needed.

(But `lib` is still included)

**Review by tejlmand — APPROVED** [review]
LGTM

**Comment by galak** [comment]
Any plans to revisit?

**Comment by mbolivar** [comment]
> Any plans to revisit?

What do you mean? This has been actively worked on for a long time. I've been sending mail to the list about it.

**Comment by mbolivar** [comment]
Also, it's a draft, so I'm not sure why you added DNM :).

**Comment by mbolivar** [comment]
kicking CI

File: doc/guides/west/manifest.rst [hunk 5]
```diff
@@ -271,3 +285,865 @@ repository root.
...
+   manifest:
+     projects:
+       - name: some-project
+         import: ...
+     self:
+       import: ...
+
+You can use this feature to import projects from elsewhere (either another git
+repository, or somewhere else inside your manifest repository) into your own
+:file:`west.yml`.
```
Comment (tejlmand) [inline doc/guides/west/manifest.rst]: I think this should be rephrased so it becomes clearer what is what.
- self import section can be used to load additional files from inside your manifest repo.
- project import can be used to load additional projects described in a git repository manifest file.

Note: should we avoid referring directly to `git` to be prepared for the day when `west` supports non-git projects ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Did an second round of reading.

Besides the given comment, I think it could help users reading about imports for the first time if the import is split clearer into two sections.
One which part of the doc describing `self::import` and another part describing `project::import` as they are quite different in their usage / purpose.
You already do describe both way, so what I'm suggesting is simply a clearer separation.

**Comment by mbolivar** [comment]
> I think it could help users reading about imports for the first time if the import is split clearer into two sections.

That sounds fine. How would you suggest organizing the example named "import order illustration", though? That combines both types.

> I think this should be rephrased so it becomes clearer what is what.

Sounds good, along the above lines.

> Note: should we avoid referring directly to git to be prepared for the day when west supports non-git projects ?

Let's not worry about that for now, I think. That day is a long way off, if it ever comes, and I'm not sure we'll allow imports from such things.

**Comment by tejlmand** [comment]
> 
> 
> > I think it could help users reading about imports for the first time if the import is split clearer into two sections.
> 
> That sounds fine. How would you suggest organizing the example named "import order illustration", though? That combines both types.

Let me try to re-organize and place a suggestion as additional commit.
Then it can be squashed or adjusted.

> 
> > I think this should be rephrased so it becomes clearer what is what.
> 
> Sounds good, along the above lines.
> 

See above.

> > Note: should we avoid referring directly to git to be prepared for the day when west supports non-git projects ?
> 
> Let's not worry about that for now, I think. That day is a long way off, if it ever comes, and I'm not sure we'll allow imports from such things.

agreed, and as I expected, hence the reason for that to be minor note.

**Comment by mbolivar** [comment]
> Let me try to re-organize and place a suggestion as additional commit.
> Then it can be squashed or adjusted.
> 

That's fine, but I'd like to merge this ASAP once we have cut west 0.7, so I would like to request that we keep it as-is unless you have a concrete proposal.

File: doc/guides/west/manifest.rst (hunk 5 above)
Comment (mbolivar) [reply doc/guides/west/manifest.rst]: > I think this should be rephrased so it becomes clearer what is what.

Addressed, thanks

**Comment by mbolivar** [comment]
Merged via https://github.com/zephyrproject-rtos/zephyr/pull/22507
